module/zfs/dmu_recv.c

```diff
@@ -2558,6 +2589,173 @@ receive_process_record(struct receive_writer_arg *rwa,
 	return (err);
 }
 
+static int
+receive_heal_record(struct receive_writer_arg *rwa,
+    struct receive_record_arg *rrd)
+{
+	uint64_t size, lsize = 0, csize = 0, blkid = 0, obj = ZB_ROOT_OBJECT;
+	dmu_buf_t *dbp;
+	abd_t *abd;
+	void *buf;
+	int err = 0;
+	blkptr_t *bp = NULL;
+	objset_t *os = rwa->os;
+	dmu_replay_record_t drr = rrd->header;
+	enum zio_flag flags = ZIO_FLAG_SPECULATIVE |
+	    ZIO_FLAG_DONT_CACHE | ZIO_FLAG_DONT_RETRY | ZIO_FLAG_CANFAIL;
+
+	/* Should only get here if we are doing a corrective recv */
+	ASSERT(rwa->heal);
+	/* Processing in order, therefore bytes_read should be increasing. */
+	ASSERT3U(rrd->bytes_read, >=, rwa->bytes_read);
+
+	rwa->bytes_read = rrd->bytes_read;
+
+	/* We can only heal write and spill records; other ones get ignored */
+	if (drr.drr_type != DRR_WRITE && drr.drr_type != DRR_SPILL)
+		goto cleanup;
+
+	switch (drr.drr_type) {
+	case DRR_WRITE:
+	{
+		uint64_t offset = drr.drr_u.drr_write.drr_offset;
+		obj = drr.drr_u.drr_write.drr_object;
+		lsize = drr.drr_u.drr_write.drr_logical_size;
+		csize = drr.drr_u.drr_write.drr_compressed_size;
+
+		buf = kmem_alloc(lsize, KM_SLEEP);
+		/* Try to read the object to see if it needs healing */
+		err = dmu_read(os, obj, offset, lsize, buf,
+		    DMU_READ_NO_PREFETCH);
+		kmem_free(buf, lsize);
+		if (err != ECKSUM)
+			goto cleanup; /* no corruption found */
+
+		err = dmu_buf_hold_noread(os, obj, offset, FTAG, &dbp);
+		if (err != 0) {
+			err = SET_ERROR(EBUSY);
+			goto cleanup;
+		}
+		blkid =
+		    dbuf_whichblock(DB_DNODE((dmu_buf_impl_t *)dbp), 0, offset);
+		/* Get the block pointer for the corrupted block */
+		bp = dmu_buf_get_blkptr(dbp);
+		dmu_buf_rele(dbp, FTAG);
+
+		break;
+	}
+	case DRR_SPILL:
+	{
+		dnode_t *dn;
+		dmu_buf_t *dbp_spill;
+		obj = drr.drr_u.drr_spill.drr_object;
+		lsize = drr.drr_u.drr_spill.drr_length;
+		csize = drr.drr_u.drr_spill.drr_compressed_size;
+
+		err = dmu_bonus_hold(os, obj, FTAG, &dbp);
+		if (err != 0)
+			goto cleanup;
+
+		/* Try to read the spill block to see if it needs healing */
+		err = dmu_spill_hold_existing(dbp, FTAG, &dbp_spill);
+		if (err != ECKSUM) {
+			dmu_buf_rele(dbp, FTAG);
+			if (err == 0)
+				dmu_buf_rele(dbp_spill, FTAG);
+			goto cleanup; /* no corruption found */
+		}
+		/* Get the block pointer to the corrupted spill block */
+		dn = dmu_buf_dnode_enter(dbp);
+		bp = DN_SPILL_BLKPTR(dn->dn_phys);
+		dmu_buf_dnode_exit(dbp);
+		dmu_buf_rele(dbp, FTAG);
+		break;
+	}
+	default:
+		ASSERT0(1);
+		err = SET_ERROR(EINVAL);
+		goto cleanup;
+	}
+
+	/* Get the good data from the recv record */
+	abd = abd_get_from_buf(rrd->arc_buf->b_data,
+	    arc_buf_size(rrd->arc_buf));
+	abd_take_ownership_of_buf(abd, B_FALSE);
+
+	if (arc_get_compression(rrd->arc_buf) != BP_GET_COMPRESS(bp)) {
+		/*
+		 * The compression in the stream doesn't match what we had
+		 * on disk; we need to re-compress the buf into the
```

Comment:
Do we really need to handle this case?  Seems like we could say that it needs to be a `zfs send --compressed` stream to use it with `zfs recv -c`.  Plus, trying to recompress it and get the exact same byte stream adds more restrictions on changing the checksum algorithm implementations, which we'd like to avoid. (cc @allanjude)